zebra-consensus/src/checkpoint.rs

```diff
@@ -3,315 +3,1098 @@
+    /// checkpoint. Also returns an error if the block or maximum heights are
+    /// missing.
+    fn check_block_height(&self, block: Arc<Block>) -> Result<BlockHeight, Error> {
+        let block_height = block
+            .coinbase_height()
+            .ok_or("the block does not have a coinbase height")?;
+        let max_checkpoint_height = self
+            .get_max_checkpoint_height()
+            .ok_or("the checkpoint list is empty")?;
+        if block_height > max_checkpoint_height {
+            return Err("the block is higher than the maximum checkpoint".into());
+        }
+        if let Some(previous_checkpoint_height) = self.get_previous_checkpoint_height() {
+            if block_height <= previous_checkpoint_height {
+                return Err("a block at this height has already been verified".into());
+            }
+        }
+        Ok(block_height)
+    }
+
+    /// Queue `block` for verification, and return `(height, hash)`.
+    ///
+    /// Verification will finish when the chain to the next checkpoint is complete,
+    /// and the caller will be notified via `tx`.
+    ///
+    /// If the block does not have a coinbase height, sends an error on `tx`, does
+    /// not queue the block, and returns None.
+    fn insert_queued_block(
+        &mut self,
+        block: Arc<Block>,
+        tx: oneshot::Sender<Result<BlockHeaderHash, Error>>,
+    ) -> Option<(BlockHeight, BlockHeaderHash)> {
+        // Check for a valid height
+        let height = match self.check_block_height(block.clone()) {
+            Ok(height) => height,
+            Err(error) => {
+                // Sending might fail, depending on what the caller does with rx,
+                // but there's nothing we can do about it.
+                let _ = tx.send(Err(error));
+                return None;
+            }
+        };
+
+        // Add the block to the list of queued blocks at this height
+        let hash = block.as_ref().into();
+        let new_qblock = QueuedBlock { block, hash, tx };
+        self.queued.entry(height).or_default().push(new_qblock);
+
+        Some((height, hash))
+    }
+
+    /// Return `(height, hash)` for the next block needed to verify the
+    /// current checkpoint range. Walks the chain of queued blocks, starting
+    /// from the next checkpoint, and following the parent hash of each block.
+    ///
+    /// If the `height` is equal to the previous checkpoint height, then the
+    /// `hash` should be equal to the previous checkpoint hash. When verifying
+    /// the genesis block, returns `(None, [0; 32])` because the genesis block
+    /// has no parent block. (And in Bitcoin, `null` is `[0; 32]`.)
+    ///
+    /// If there are no queued blocks for the current range, returns the next
+    /// checkpoint height and hash. If checkpoint verification has finished,
+    /// returns `(None, None)`.
+    fn find_checkpoint_chain_end(&self) -> (Option<BlockHeight>, Option<BlockHeaderHash>) {
```

Comment:
Should this be an `Option<(Height, Hash)>`?

Reply:
I think it should probably be an enum, rather than requiring the user to parse various combinations of "None" and "Some".

Reply:
I'll try to fix this during the enum refinement in https://github.com/ZcashFoundation/zebra/pull/524#discussion_r448098783

Reply:
Turned into an enum in 272a9b27bd59930858bf2ff056b82d2e29137524 as part of the algorithm rewrite.